### ISSUE TITLE
[11.0] Unable to install base_setup on Ubuntu 18.04, python 3.6, lxml 3.7.1

### ISSUE THREAD

**Issue by dalonsod** [odoo/odoo#29556]
**_Impacted versions_**: 
Branch 11.0, tested on commit a6d13535916, Ubuntu 18.04, python 3.6 virtualenv 

**_Steps to reproduce_**: 
From the related commit, clone Odoo and install python requirements, that will install lxml on 3.7.1 version, as expected (file odoo/requirements.txt):

```
lxml==3.7.1 ; sys_platform != 'win32' and python_version < '3.7'
lxml==4.2.3 ; sys_platform != 'win32' and python_version >= '3.7'

```
Then, create and empty database running Odoo without demo data.

**_Current behavior_**:
When trying to install any module from Applications menu, an error is thrown. In particular, when installing base_setup addon the trace output generated is shown below:

```
ERROR conta_campelo odoo.modules.registry: Failed to load registry
Traceback (most recent call last):
  File "/home/odoo/odoo-prod/conta-campelo/src/odoo/odoo/modules/registry.py", line 85, in new
    odoo.modules.load_modules(registry._db, force_demo, status, update_module)
  File "/home/odoo/odoo-prod/conta-campelo/src/odoo/odoo/modules/loading.py", line 380, in load_modules
    loaded_modules, update_module, models_to_check)
  File "/home/odoo/odoo-prod/conta-campelo/src/odoo/odoo/modules/loading.py", line 274, in load_marked_modules
    perform_checks=perform_checks, models_to_check=models_to_check
  File "/home/odoo/odoo-prod/conta-campelo/src/odoo/odoo/modules/loading.py", line 181, in load_module_graph
    _load_data(cr, module_name, idref, mode, kind='data')
  File "/home/odoo/odoo-prod/conta-campelo/src/odoo/odoo/modules/loading.py", line 95, in _load_data
    tools.convert_file(cr, module_name, filename, idref, mode, noupdate, kind, report)
  File "/home/odoo/odoo-prod/conta-campelo/src/odoo/odoo/tools/convert.py", line 791, in convert_file
    convert_xml_import(cr, module, fp, idref, mode, noupdate, report)
  File "/home/odoo/odoo-prod/conta-campelo/src/odoo/odoo/tools/convert.py", line 837, in convert_xml_import
    etree.parse(os.path.join(config['root_path'],'import_xml.rng' )))
  File "src/lxml/lxml.etree.pyx", line 3427, in lxml.etree.parse (src/lxml/lxml.etree.c:81110)
  File "src/lxml/parser.pxi", line 1811, in lxml.etree._parseDocument (src/lxml/lxml.etree.c:117841)
  File "src/lxml/parser.pxi", line 1837, in lxml.etree._parseDocumentFromURL (src/lxml/lxml.etree.c:118188)
  File "src/lxml/parser.pxi", line 1741, in lxml.etree._parseDocFromFile (src/lxml/lxml.etree.c:117100)
  File "src/lxml/parser.pxi", line 1138, in lxml.etree._BaseParser._parseDocFromFile (src/lxml/lxml.etree.c:111646)
  File "src/lxml/parser.pxi", line 595, in lxml.etree._ParserContext._handleParseResultDoc (src/lxml/lxml.etree.c:105102)
  File "src/lxml/parser.pxi", line 706, in lxml.etree._handleParseResult (src/lxml/lxml.etree.c:106810)
  File "src/lxml/parser.pxi", line 646, in lxml.etree._raiseParseError (src/lxml/lxml.etree.c:105956)
  File "b'/home/odoo/odoo-prod/conta-campelo/src/odoo/odoo/import_xml.rng'", line 0
lxml.etree.XMLSyntaxError: <no detail available>
```

This error appears to link to a problem with XML parser (lxml) and an unknown XML file of the related addon. **Then, a simple python lxml package upgrade through pip solves this problem** (tested with lxml==4.2.3).

Other approach to solve this problem, without upgrading lxml, was this one (modify odoo/tools/convert.py in the same way):
https://github.com/odoo/odoo/issues/13815#issuecomment-367210431

**_Expected behavior_**:
Once the database is created, any Odoo builtin addon is expected to be installable without additional stuff.

**Comment by pedrobaeza** [comment]
The official supported Python version is 3.5. Have you tried with that version?

**Comment by dalonsod** [comment]
OK, I'll try as soon as possible.

[Official documentation](https://www.odoo.com/documentation/11.0/setup/install.html#id3) refers to "Python 3.5+", so it appears to be incorrect...

**Comment by pedrobaeza** [comment]
Yes, maybe that `+` is too optimistic... That's why I want you to try that first.

**Comment by moylop260** [comment]
The runbot's odoo uses python3.6 then you shouldn't be this problem.
Maybe you are installing a environment and using another one.

if you uses `odoo-bin` file uses `python3` magic comment to run it.

Then try using the same command to install:
`sudo python3 -m pip install --force-reinstall -U -r requirements.txt`

Maybe is the pip version, you can try updating it:
`sudo python3 -m pip install -U --force-reinstall pip`

If you are using a `virtualenv` then:
skip `sudo`
skip `python3` (use just python)
Starting server using explicit python command: `python odoo-bin`

**Comment by probabble** [comment]
Getting the same error with Odoo 12, Ubuntu 18.04, system python3 env (Docker).

**Comment by probabble** [comment]
Also, confirmed that upgrading to `lxml==4.2.3` fixes the error.

**Comment by moylop260** [comment]
Could you create a proper PR for requirements.txt file changing:
```patch
- lxml==4.2.3 ; sys_platform != 'win32' and python_version >= '3.7'
lxml==4.2.3 ; sys_platform != 'win32' and python_version >= '3.6'
```
please?

**Comment by moylop260** [comment]
@probabble 
Could you share me the smaller Dockerfile script to reproduce it, please?

**Comment by moylop260** [comment]
FYI I have created the following Dockerfile content:

```Dockerfile
FROM ubuntu:18.04

RUN apt-get update && apt-get install -y ca-certificates \
    curl libssl1.0-dev node-less xz-utils git \
    python3 python3-pip libxml2-dev libxslt-dev libevent-dev \
    libsasl2-dev libldap2-dev postgresql-client
RUN git clone --depth=1 --single-branch https://github.com/odoo/odoo.git -b 12.0
RUN python3 -m pip install -Ur odoo/requirements.txt
```

Then I just build:
`docker build -t odoo-ubuntu18-py36 .`

Create a container for postgresql:
`docker run -d -e POSTGRES_USER=root -e POSTGRES_PASSWORD=root -e POSTGRES_DB=postgres --name db postgres:10`

Create a odoo container and linked to database postgresql container
`docker run -it -p 8069:8069 --name odoo_ubuntu_18_py36 --link db:db -t odoo-ubuntu18-py36 bash`

I assigned the PG environment variables from odoo container to linked container.
```bash
export PGHOST=${DB_PORT_5432_TCP_ADDR}
export PGPORT=${DB_PORT_5432_TCP_PORT}
export PGPASSWORD=${DB_ENV_POSTGRES_PASSWORD}
```

Start odoo server:
`/odoo/odoo-bin -d borrar`

Open a browser to connect to odoo and install `base_setup` module.

Result:
- ![module_installing](https://user-images.githubusercontent.com/6644187/50193868-18a62080-02fe-11e9-9865-e39280fe0c1d.gif)

I mean, I can't reproduce the error.
If you share us your Dockerfile in order to reproduce it we can help you better.

**Comment by dalonsod** [comment]
> The runbot's odoo uses python3.6 then you shouldn't see this issue.
> Maybe you are installing a environment and using another one.
> 
> if you uses `odoo-bin` file uses `python3` magic comment to run it.
> 
> Then try using the same command to install:
> `sudo python3 -m pip install --force-reinstall -U -r requirements.txt`
> 
> Maybe is the pip version, you can try updating it:
> `sudo python3 -m pip install -U --force-reinstall pip`
> 
> If you are using a `virtualenv` then:
> skip `sudo`
> skip `python3` (use just python)
> Starting server using explicit python command: `python odoo-bin`

@moylop260 thanks for your comments, the installation was made with a Python 3.6 virtualenv, no sudoer user. I'm using Odoo 11 & Ubuntu 18.04 & Python 3.6 virtual environments since 2018 Q2 without any problem, I think a later version Odoo 11 made the difference. However, I'll try to upgrade PIP before downgrading Python version as @pedrobaeza recommended

**Comment by dalonsod** [comment]
@moylop260 got the same issue with pip 18.1
@pedrobaeza with Python 3.5(.6) works perfectly! As you said, "3.5+" has been too optimistic

**Comment by probabble** [comment]
I hit this issue attempting to upgrade a module from OCA. I can reproduce the error when upgradig `base_setup`. Try `odoo-bin -u base_setup [options]`.

**Comment by moylop260** [comment]
I just changed the `Dockerfile` in order to get the same sha reported.

And the result was the following:
```Dockerfile
FROM ubuntu:18.04

RUN apt-get update && apt-get install -y ca-certificates \
    curl libssl1.0-dev node-less xz-utils git \
    python3 python3-pip libxml2-dev libxslt-dev libevent-dev \
    libsasl2-dev libldap2-dev postgresql-client
RUN git clone https://github.com/odoo/odoo.git -b 11.0 && \
    git --git-dir=/odoo/.git --work-tree=/odoo reset --hard a6d1353
RUN python3 -m pip install -Ur /odoo/requirements.txt
```

Then I just created a new empty database without demo installing `base_setup` module.
```bash
root@aa4521538ab4:/# /odoo/odoo-bin --without-demo=all -d lxml_error -i base_setup --stop-after-init --log-level=error
Running as user 'root' is a security risk.

root@aa4521538ab4:/# echo $?
0

root@aa4521538ab4:/# cat /etc/lsb-release
DISTRIB_ID=Ubuntu
DISTRIB_RELEASE=18.04
DISTRIB_CODENAME=bionic
DISTRIB_DESCRIPTION="Ubuntu 18.04.1 LTS"

root@aa4521538ab4:/# python3 --version
Python 3.6.7

root@aa4521538ab4:/# python3 -m pip freeze | grep lxml
lxml==3.7.1

root@aa4521538ab4:/# python3 -c "import lxml.etree;print(lxml.etree.__version__)"
3.7.1

root@aa4521538ab4:/# python3 -c "import lxml;print(lxml.__file__)"
/usr/local/lib/python3.6/dist-packages/lxml/__init__.py

root@aa4521538ab4:/# git --git-dir=/odoo/.git --work-tree=/odoo log -r -1 --oneline
b750c78b (grafted, HEAD -> 12.0, origin/HEAD, origin/12.0) [FIX] web: translate terms on search view
```

I'm using a similar environment like you description but it is not reproduced it yet.
Runbot is using python3.6 and the builds are green.

Using virtualenv you need check if system-packages is enabled.
If you have installed from apt-get the python-lxml package that is generating conflicts.
If the main python (not virtual env) has installed another lxml package version.

Could you share us the same output of the commands above, please?
Or better, send us the Dockerfile script to generate the real case with the error.

Maybe you have another additional package that is generating a lxml corrupted.
I don't know, but if we can't reproduce it then we can not fix it.

Thanks in advance.

**Comment by moylop260** [comment]
Seems like a error with a particular xml.
Could you apply the following patch, in order to detect the xml with the problem?

```patch
diff --git a/odoo/tools/convert.py b/odoo/tools/convert.py
index b98e586a..c6a3fe1c 100644
--- a/odoo/tools/convert.py
+++ b/odoo/tools/convert.py
@@ -788,6 +788,7 @@ form: module.record_id""" % (xml_id,)
         }

 def convert_file(cr, module, filename, idref, mode='update', noupdate=False, kind=None, report=None, pathname=None):
+    _logger.warning("Converting file %s/%s", module, filename)
     if pathname is None:
         pathname = os.path.join(module, filename)
     ext = os.path.splitext(filename)[1].lower()
```

And share us the output, please?

**Comment by probabble** [comment]
@moylop260 I'm on a tight deadline to get Odoo deployed, so haven't dived into this. Short story: I can't get reliable replication of this issue from clean Docker installations. 

That _doesn't mean it's not a bug_. Even if the source of the issue is some module that loads corrupted XML, Odoo should be detecting that the XML isn't okay before loading it and triggering this error.

For now, I'm just installing the higher version of `lxml` as @dalonsod suggested. Way easier than picking apart the exact combination of configurations that led to the error.

**Comment by fgarcia-humanoide** [comment]
I have the same issue in an odoo 11 ubuntu 18.04

File "b'/opt/odoo/odoo-server/odoo/import_xml.rng'", line 0
lxml.etree.XMLSyntaxError:

Can not install modules
Can not update modules

Two days ago installed 120 modules.

Now i can't

**Comment by Yenthe666** [comment]
Hi @dalonsod,

Where you able to solve this issue eventually or is it still relevant?

**Comment by dalonsod** [comment]
@Yenthe666 from my point of view the original issue is solved, using Python 3.5 instead of 3.6 one.

**Comment by Yenthe666** [comment]
Okay thanks, I'll close it for now anyways.
